### DISCUSSION TITLE
Chapter4 Task 8 (A conundrum)

### DISCUSSION THREAD

**Discussion by sa-mendez** [kowainik/learn4haskell#403] [General]
First of all, I love the course.  I have been dabbling in Haskell over the last few months when I stumbled upon your course, and really appreciate it as a resource.

The Task is as follows:

```
Can you implement a monad version of AND, polymorphic over any monad?
🕯 HINT: Use "(>>=)", "pure" and anonymous function
-}
andM :: (Monad m) => m Bool -> m Bool -> m Bool
```

I went ahead and implemented it (wont post solution here) as prescribed, and upon running the tests, only the following test case failed:

`it "Just False - Nothing" $ andM (Just False) Nothing `shouldBe` Just False`

Intuitively, this makes sense, because of boolean short circuiting.  However, if you run the following in ghci (note that this would be equivalent to a cheating implementation of  andM = liftM2 (&&)  ):

liftM2 (&&) (Just False) Nothing

You get Nothing, which would also fail the unit test, and is at least consistent with what my implementation outputs.

I did update my implementation to short-circuit and got the test to pass, but given the results of using liftM2, is short-circuiting necessary for correctness when doing boolean AND in a monadic context?  

Thank in advance,

Sergio

**Comment by vrom911** [comment]
The ordinary `&&` is lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face: 


If you don't use the `>>=` operator or use `liftM2` (that have the same behaviour as `liftA2` unfortunately), then the `Monad` constraint is too strong, and it is enough to specify only `Applicative`. However, with `Applicative` you are not able to implement a lazier version of `andM`.

**Reply by sa-mendez** [reply]
Thank you for clearing that up.

Out of curiosity I went and implemented andM using do notation, and my naive implementation of it suffered from the same problem where it was not lazy on the second argument.  I had to purposely introduce short-circuiting in my do notation solution in order for 

`Just False `andM` _ `

to properly yield Just False
